### ISSUE TITLE
Provide a way to open a browser for server/deployment

### ISSUE THREAD

**Issue by lstocchi** [redhat-developer/vscode-adapters#299]
It would be useful to add a button, inside the context menu, which allows to open a browser pointing at the server/deployment started, or maybe open the browser automatically when the server is started.

**Comment by robstryker** [comment]
It turns out that determining the URL of a given webapp is actually quite difficult ;)  For the simplest case it's probably just localhost:8080/warName   but that's not always true.

**Comment by lstocchi** [comment]
@robstryker I still have to check part of rsp-server code so maybe what i'm going to ask you is no sense but ... isn't the port 8080 always used by default because it is written in the code? 

1. Wouldn't it be better to use a port randomly chosen? 
2. Even though the port used is 8080 i'd still need a confirmation from the server side before opening a browser page or display an error message like "the port is already used". What do you suggest to do?

**Comment by robstryker** [comment]
To make this work for any and all server adapter types, it'd actually need to be an API provided by the server.  Let's walk through this. 

The rsp-server is more of a framework. It provides an API for clients to communicate to the server. The server can be extended by additional osgi bundles. But the server could theoretically hold ANY type of server:  felix, karaf, tomcat, jboss, wildfly, eap, minishift, the list could go on and on and on. 

So if we accept that almost ANY type of server can be created with the rsp-server framework, then there must be some servers that have hard-coded ports other than 8080. Or, there must be some servers that are configured to use a port via a sysprop. Or maybe even a text .properties file. So really only the "server adapter" (ie the rsp-server extension in charge of that type of server (jboss, wildfly, glassfish, eap, minishift, etc)  would know what files to check to discover the port). 

Now, what about a full url? Other than host+port, it also requires the rest of a url. While jboss / wildfly / eap / and others will check the deployment's context root, other types of servers might not. Some deployments might not have a URL at all (imagine a simple osgi bundle being deployed to apache tomcat). 

So, only the server type itself would know where to find the host, the port, or the rest of the url. Only the server type itself would know if a given deployment even HAS a url to open. 

Therefore, discovering the "url" must be an operation made to the rsp-server. The rsp-server will then pass the request on to the given server adapter type (wildfly, eap, glassfish, felix, etc), and ask it "Does this module have a url? If yes, what is it?"

So basically, vscode-adapters should *NOT* make *ANY* code to assume a url for a given module. This would need to be a rsp-server API.

**Comment by robstryker** [comment]
> imagine a simple osgi bundle being deployed to apache tomca

I mean apache felix. Oops

**Comment by lstocchi** [comment]
I agree with your final thought, maybe i should have opened an issue on rsp-server project, i was referring to that one (should we close this issue and/or move it to the rsp-server?)... what i was missing was the complete view about this project. Now it's clearer your "quite difficult" statement 👍

**Comment by adietish** [comment]
I'd back the need for this but I also ack the required work for it to be possible. Nevertheless I deem this important for the user experience and thus suggest that we start tackling it in the 166 milestone.

**Comment by adietish** [comment]
* first step to fulfill this is to be able to edit servers: https://github.com/redhat-developer/rsp-server/pull/386
* then we have to specifically allow editing the host or port of a wildfly server: https://github.com/redhat-developer/rsp-server/issues/393

Moving this issue into the next sprint for the further steps to be taken to be able to provide a solution for this issue.
